gobblin-utility/src/main/java/gobblin/util/request_allocation/ResourcePool.java

```diff
@@ -0,0 +1,124 @@
+
+    int currentIndex = 0;
+    for (Map.Entry<String, Double> resource : maxResources.entrySet()) {
+      indexBuilder.put(resource.getKey(), currentIndex);
+      this.softBound[currentIndex] = resource.getValue();
+      currentIndex++;
+    }
+    this.dimensionIndex = indexBuilder.build();
+
+    this.hardBound = this.softBound.clone();
+    for (int i = 0; i < this.hardBound.length; i++) {
+      this.hardBound[i] *= DEFAULT_DIMENSION_TOLERANCE;
+    }
+    this.defaultResourceUse = new double[this.softBound.length];
+
+    for (Map.Entry<String, Integer> idxEntry : this.dimensionIndex.entrySet()) {
+      if (tolerances.containsKey(idxEntry.getKey())) {
+        this.hardBound[idxEntry.getValue()] =
+            this.softBound[idxEntry.getValue()] * Math.max(1.0, tolerances.get(idxEntry.getKey()));
+      }
+      if (defaultRequirements.containsKey(idxEntry.getKey())) {
+        this.defaultResourceUse[idxEntry.getValue()] = defaultRequirements.get(idxEntry.getKey());
+      }
+    }
+  }
+
+  private ResourcePool(double[] softBound, double[] hardBound, double[] defaultResourceUse, ImmutableMap<String, Integer> dimensionIndex) {
+    this.softBound = softBound;
+    this.hardBound = hardBound;
+    this.defaultResourceUse = defaultResourceUse;
+    this.dimensionIndex = dimensionIndex;
+  }
+
+  /**
+   * @return true if input {@link ResourceRequirement} exceeds the soft bound long any dimension. If the parameter
+   *        orEqual is true, then matching along any dimension will also return true.
+   */
+  public boolean exceedsSoftBound(ResourceRequirement requirement, boolean orEqual) {
+    return VectorAlgebra.exceedsVector(this.softBound, requirement.getResourceVector(), orEqual);
+  }
+
+  /**
+   * @return true if input {@link ResourceRequirement} exceeds the hard bound long any dimension. If the parameter
+   *        orEqual is true, then matching along any dimension will also return true.
+   */
+  public boolean exceedsHardBound(ResourceRequirement requirement, boolean orEqual) {
+    return VectorAlgebra.exceedsVector(this.hardBound, requirement.getResourceVector(), orEqual);
+  }
+
+  /**
+   * Use to create a {@link ResourceRequirement} compatible with this {@link ResourcePool}.
+   */
+  public ResourceRequirement.Builder getResourceRequirementBuilder() {
+    return new ResourceRequirement.Builder(this);
+  }
+
+  ResourcePool contractPool(ResourceRequirement requirement) {
+    return new ResourcePool(VectorAlgebra.addVector(this.softBound, requirement.getResourceVector(), -1., null),
+        VectorAlgebra.addVector(this.hardBound, requirement.getResourceVector(), -1., null),
+        this.defaultResourceUse, this.dimensionIndex);
+  }
+
+  int getNumDimensions() {
+    return this.dimensionIndex.size();
+  }
+
+  double[] getDefaultResourceUse(double[] reuse) {
+    return reuseOrClone(this.defaultResourceUse, reuse);
+  }
+
+  private double[] reuseOrClone(double[] ref, double[] reuse) {
```

Comment:
Can you move this method to `getDefaultResourceUse`? It is only called once, and would be more clear.